pkg/cli/mirror/copy.go

```diff
@@ -110,70 +121,93 @@ func (o *MirrorOptions) bulkImageCopy(isc *v1alpha2.ImageSetConfiguration, srcSk
-							Category:            v1alpha2.TypeOperatorRelatedImage,
-						}
-						dstTIR, err := image.ParseReference("file://" + file.Name() + "/" + name)
-						if err != nil {
-							return err
-						}
-						dstTI := image.TypedImage{
-							TypedImageReference: dstTIR,
-							Category:            v1alpha2.TypeOperatorRelatedImage,
-						}
-						mapping[srcTI] = dstTI
+					srcTI := image.TypedImage{
+						TypedImageReference: srcTIR,
+						Category:            v1alpha2.TypeOperatorRelatedImage,
+					} //pkg.Name + "/" + folder
+					dstPath := "file://" + pkg.Name + "/" + name
+					if srcTIR.Ref.ID != "" {
+						dstPath = dstPath + "/" + strings.TrimPrefix(srcTI.Ref.ID, "sha256:")
+					} else if srcTIR.Ref.ID == "" && srcTIR.Ref.Tag != "" {
+						dstPath = dstPath + "/" + fmt.Sprintf("%x", sha256.Sum256([]byte(srcTIR.Ref.Tag)))[0:6]
+					}
+					dstTIR, err := image.ParseReference(dstPath)
+					if err != nil {
+						return err
 					}
-					break
+					if srcTI.Ref.Tag != "" {
+						dstTIR.Ref.Tag = srcTI.Ref.Tag
+					}
+					dstTI := image.TypedImage{
+						TypedImageReference: dstTIR,
+						Category:            v1alpha2.TypeOperatorRelatedImage,
+					}
+					mapping[srcTI] = dstTI
 				}
+				break
 			}
 		}
 	}
 
 	o.Dir = strings.TrimPrefix(o.Dir, o.OutputDir+"/")
-	err := remoteRegFuncs.mirrorMappings(*isc, mapping, srcSkipTLS)
-	if err != nil {
-		return err
+	if len(mapping) > 0 {
+		err := remoteRegFuncs.mirrorMappings(*isc, mapping, srcSkipTLS)
+		if err != nil {
+			return err
+		}
+	} else {
+		klog.Infof("no images to copy")
 	}
 
 	return nil
 }
 
 // bulkImageMirror used to mirror the relevant images (push from a directory) to
 // a remote registry in oci format
-func (o *MirrorOptions) bulkImageMirror(isc *v1alpha2.ImageSetConfiguration, destRepo, namespace string, srcSkipTLS, dstSkipTLS bool, insecureSigPolicy bool, remoteRegFuncs RemoteRegFuncs) error {
+func (o *MirrorOptions) bulkImageMirror(isc *v1alpha2.ImageSetConfiguration, destRepo, namespace string, remoteRegFuncs RemoteRegFuncs) error {
 	mapping := image.TypedImageMapping{}
+
 	for _, operator := range isc.Mirror.Operators {
 		_, _, repo, _, _ := parseImageName(operator.Catalog)
-		log.Printf("INFO: processing contents of local catalog %s\n", operator.Catalog)
+
+		// Mirroring workflow assumes a Copy workflow has been executed before.
+		// In this case, the copy workflow has untarred the contents of the catalog to
+		// an artifacts subfolder under the destination directory previously set for the copy.
+		// Example:
+		// Copy workflow used oci://mydir as the destination folder to copy catalog mycatalog.
+		// The OCI image for mycatalog will be in mydir/mycatalog.
+		// The untarred contents of mycatalog will be in mydir/olm_artifacts/operator1..n
+		// Here we reconstruct the artifacts subfolder for each operator for the operator oci://mydir/mycatalog
+		// artifactsPath represents in this example mydir/olm_artifacts
```

Comment:
The overall logic here assumes that the `olm_artifacts` directory has already been created. As the docs state, it's assuming that the copy function has been executed. I don't think this is an entirely valid assumption. The copy function is certainly a convenient way to get an OCI catalog image on the file system, but that's not the use case we need to satisfy. We will be generating our own docker/OCI image on the file system, using our own tooling, so our users would never really use the "copy command" (it's great for testing though).

The idea was to provide a `oci://` reference to oc mirror through the ImageSetConfiguration file (and this reference could be pointing anywhere on the file system), and then it's up to oc mirror to figure out how to fetch the OLM artifacts from the catalog (and if necessary store them somewhere, in this case inside `olm_artifacts`). 

The point I am trying to make here is that I don't think the assumption should be that `olm_artifacts` exists. The code should check to see if it is already present and handle this accordingly:

- if `olm_artifacts/<operator>` is present and has has the operator we are looking for, use it as-is
- if `olm_artifacts/<operator>` is not present, oc mirror should do what the copy function does and creates that directory and populates it.

Let me know if I need to clarify anything here.

Reply:
Here's what we propose:
For this TechPreview, your users will still need to call (as an example)
* `./bin/oc-mirror --config imstcfg_preparation.yaml --use-oci-feature --oci-feature-action=copy oci://a-directory-where-oc-mirror-sets-up-stuff`
  with ImageSetConfig:
```yaml
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v1alpha2
mirror:
  operators:
  - catalog: oci:///home/skhoury/go/src/github.com/openshift/oc-mirror/wrklds-547/redhat-operator-index
    packages:
    - name: node-observability-operator
    - name: aws-load-balancer-operator
```
*  `./bin/oc-mirror --config imstcfg_mirroring.yaml --use-oci-feature --oci-feature-action=mirror docker://localhost:5000/someplace  --dest-skip-tls`
  with ImageSetConfig:
```yaml
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v1alpha2
mirror:
  operators:
  - catalog: oci://a-directory-where-oc-mirror-sets-up-stuf
    packages:
    - name: node-observability-operator
    - name: aws-load-balancer-operator
```
For the 4.13 version, it was planned to blend this feature better with the oc-mirror existing flows, and we can work on making this less cumbersome for the IBM users

Reply:
Found some slight differences in the instructions I had to use. Basically the mirror step requires the ImageSetConfiguration to contain the path to the catalog (see step 2 comment in the YAML).

- Copy from a pre-existing OCI image elsewhere on the file system to a dir where oc-mirror "sets stuff up" (i.e. `tmp/wrklds-536` in this example)

  ```bash
  oc-mirror -c '/Users/jhunkins/go/src/github.com/openshift/oc-mirror/ImageSetConfiguration.yml' --use-oci-feature --oci-feature-action copy oci:///tmp/wrklds-536/
  ```
  
  ```yaml
  kind: ImageSetConfiguration
  apiVersion: mirror.openshift.io/v1alpha2
  storageConfig:
    local:
      path: /tmp/localstorage
  mirror:
    operators:
    # step 1: use the OCI image in a fake copy operation
    - catalog: oci:///tmp/ocicontent/common-services
      packages:
      - name: ibm-auditlogging-operator-app
        channels:
        - name: v3
  additionalImages:
  - name: registry.redhat.io/ubi8/ubi:latest
  ```

- Mirror from the place where oc-mirror "sets stuff up" (i.e. `tmp/wrklds-536/` in this example) to a final registry
  
  ```bash
  oc-mirror -c '/Users/jhunkins/go/src/github.com/openshift/oc-mirror/ImageSetConfiguration.yml' --use-oci-feature --oci-feature-action mirror --oci-insecure-signature-policy --dest-skip-tls docker://localhost:5000/tmp/wrklds-536
  ```
  
  ```yaml
  kind: ImageSetConfiguration
  apiVersion: mirror.openshift.io/v1alpha2
  storageConfig:
    local:
      path: /tmp/localstorage
  mirror:
    operators:
    # step 2: mirror from the previous copy folder (NOTE: need to specify the "common-services" folder here
    - catalog: oci:///tmp/wrklds-536/common-services
      packages:
      - name: ibm-auditlogging-operator-app
        channels:
        - name: v3
  additionalImages:
  - name: registry.redhat.io/ubi8/ubi:latest
  ```

Reply:
I am including the results I got when mirroring below:

```console
$ curl --no-progress-meter http://localhost:5000/v2/_catalog | jq
{
  "repositories": [
    "tmp/wrklds-536/common-services",
    "tmp/wrklds-536/cpopen/cpfs/audit-garbage-collector",
    "tmp/wrklds-536/cpopen/cpfs/audit-policy-controller",
    "tmp/wrklds-536/cpopen/cpfs/fluentd",
    "tmp/wrklds-536/cpopen/ibm-auditlogging-operator",
    "tmp/wrklds-536/cpopen/ibm-auditlogging-operator-bundle",
    "tmp/wrklds-536/opencloudio/audit-garbage-collector",
    "tmp/wrklds-536/opencloudio/audit-policy-controller",
    "tmp/wrklds-536/opencloudio/fluentd",
    "tmp/wrklds-536/opencloudio/ibm-auditlogging-operator",
    "tmp/wrklds-536/opencloudio/ibm-auditlogging-operator-bundle"
  ]
}
```

Couple things I noticed:
- the `common-services` catalog is not in an "org"... I presume I'd need to do _something_ to make an "org" (e.g `cpopen/cpfs` or `cpopen`) but not sure what that would be
- I see that the additional images are not brought forward from the ImageSetConfiguraiton. Seems like this should happen by default if `additionalImages` are present.

Reply:
For the instructions, I'm sorry for the mistakes. But you got it right from what I see in your comment. :+1: 

Regarding additionalImages, from the initial conversations, we understood that this wasn't part of the work for 4.12 [See here](https://docs.google.com/document/d/1-TESqErOjxxWVPCbhQUfnT3XezG2898fEREuhGena5Q/edit?disco=AAAAdJ2f7cA). So I'm sure when we blend the feature back in with oc-mirror for the next release, additionalImages should fold back to the regular oc-mirror workflow. 

As for the org on the catalog, I think I see what you mean, but I don't see where to get an org from when we start from an FBC catalog...